File: backend/app/services/nlp/pipeline.py
from datetime import datetime
from typing import Dict, Any

from .section_detector import detect_sections
from .skill_extractor import extract_skills
from .entity_extractor import extract_education, extract_experiences, extract_projects, extract_certifications
from .normalizer import normalize_profile

def process_resume_nlp(raw_text: str) -> Dict[str, Any]:
    if not raw_text or not raw_text.strip():
        return {
            "summary": "",
            "roles": [],
            "skills": [],
            "education": [],
            "experience": [],
            "projects": [],
            "certifications": [],
            "analyzer_version": "nlp-v1",
            "analyzed_at": datetime.utcnow().isoformat(),
            "processing_status": "failed",
            "processing_error": "Empty text provided"
        }

    sections = detect_sections(raw_text)
    skills = extract_skills(sections, raw_text)
    education = extract_education(sections, raw_text)
    experience = extract_experiences(sections, raw_text)
    projects = extract_projects(sections, raw_text)
    certifications = extract_certifications(sections, raw_text)

    summary = sections.get("summary", "")

    result = normalize_profile({
        "summary": summary,
        "roles": [],
        "skills": skills,
        "education": education,
        "experience": experience,
        "projects": projects,
        "certifications": certifications
    })
    result["analyzed_at"] = datetime.utcnow().isoformat()
    return result

File: server/nlp/pipeline.ts
import { detectResumeSections, SectionMap } from "./sectionDetector.js";
import { extractSkillsFromResume, ExtractedSkill } from "./skillExtractor.js";
import {
  extractEducation,
  extractExperiences,
  extractProjects,
  extractCertifications,
  extractRoles,
  extractSummary,
  ExtractedEducation,
  ExtractedExperience,
  ExtractedProject,
  ExtractedCertification
} from "./entityExtractor.js";

export interface CandidateNlpProfile {
  summary: string;
  roles: string[];
  skills: ExtractedSkill[];
  education: ExtractedEducation[];
  experience: ExtractedExperience[];
  projects: ExtractedProject[];
  certifications: ExtractedCertification[];
  analyzer_version: string;
  analyzed_at: string;
  processing_status: "completed" | "failed";
  processing_error?: string;
}

/**
 * Executes the full modular NLP processing pipeline on raw resume text.
 * Stage 1: Text Cleaning & Section Detection
 * Stage 2: Skill Extraction & Normalization
 * Stage 3: Education & Experience Extraction
 * Stage 4: Project & Certification Extraction
 * Stage 5: Role & Summary Synthesis
 */
export function analyzeResumeText(rawText: string): CandidateNlpProfile {
  const now = new Date().toISOString();

  if (!rawText || !rawText.trim()) {
    return {
      summary: "",
      roles: [],
      skills: [],
      education: [],
      experience: [],
      projects: [],
      certifications: [],
      analyzer_version: "nlp-v1",
      analyzed_at: now,
      processing_status: "failed",
      processing_error: "Empty or machine-unreadable document text provided."
    };
  }

  try {
    // 1. Detect Sections
    const sections: SectionMap = detectResumeSections(rawText);

    // 2. Extract Skills
    const skills = extractSkillsFromResume(sections);

    // 3. Extract Entities
    const education = extractEducation(sections);
    const experience = extractExperiences(sections);
    const projects = extractProjects(sections);
    const certifications = extractCertifications(sections);
    const roles = extractRoles(sections);
    const summary = extractSummary(sections);

    return {
      summary,
      roles,
      skills,
      education,
      experience,
      projects,
      certifications,
      analyzer_version: "nlp-v1",
      analyzed_at: now,
      processing_status: "completed"
    };
  } catch (err: any) {
    return {
      summary: "",
      roles: [],
      skills: [],
      education: [],
      experience: [],
      projects: [],
      certifications: [],
      analyzer_version: "nlp-v1",
      analyzed_at: now,
      processing_status: "failed",
      processing_error: `NLP Analysis failed: ${err.message || "Pipeline execution error."}`
    };
  }
}
